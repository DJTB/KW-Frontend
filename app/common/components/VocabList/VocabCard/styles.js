import styled from 'styled-components';
import { mix } from 'polished';
import { gutter } from 'common/styles/layout';
import { borderRadius } from 'common/styles/sizing';

import A from 'common/components/A';

export const Wrapper = styled.li`
  display: inline-flex;
  flex: 1 0 auto;
  justify-content: center;
  color: ${({ textColor }) => textColor};
  background-color: ${({ bgColor }) => mix(0.8, bgColor, '#bbb')};
  text-shadow: 0 2px 2px ${({ bgColor }) => mix(0.8, bgColor, '#444')};
  border-radius: ${borderRadius};
  box-shadow: 2px 2px 0 rgba(0, 0, 0, 0.1);
`;

export const Link = styled(A)`
  ${gutter({ position: 'vertical', mod: 2 })}
  ${gutter({ position: 'horizontal', mod: 4 })}
  display: flex;
  width: 100%;
  flex-flow: column nowrap;
  text-align: center;
  justify-content: center;
  align-content: center;
  align-items: center;
`;
