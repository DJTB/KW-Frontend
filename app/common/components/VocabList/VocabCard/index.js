import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import VocabWord from 'common/components/VocabWord';
import VocabMeaning from 'common/components/VocabMeaning';
import getSrsRankName from 'common/utils/getSrsRankName';
import { selectStreak, selectIsHidden, selectPrimaryVocabId } from 'features/reviews/selectors';
import { white, purple, grey, SRS_COLORS } from 'common/styles/colors';

import { Wrapper, Link } from './styles';

VocabCard.propTypes = {
  id: PropTypes.number.isRequired,
  vocabId: PropTypes.number.isRequired,
  showSecondary: PropTypes.bool,
  showFuri: PropTypes.bool,
  bgColor: PropTypes.string,
  textColor: PropTypes.string,
};

VocabCard.defaultProps = {
  showSecondary: false,
  showFuri: false,
  bgColor: purple[5],
  textColor: white[1],
};

export function VocabCard({ id, vocabId, showFuri, showSecondary, textColor, bgColor, ...props }) {
  return (
    <Wrapper textColor={textColor} bgColor={bgColor} {...props}>
      <Link plainLink to={`/vocabulary/entry/${id}`}>
        <VocabWord id={vocabId} showFuri={showFuri} showSecondary={showSecondary} />
        <VocabMeaning id={id} showSecondary={showSecondary} />
      </Link>
    </Wrapper>
  );
}

const mapStateToProps = (state, props) => {
  const streak = selectStreak(state, props);
  const isHidden = selectIsHidden(state, props);
  const vocabId = selectPrimaryVocabId(state, props);
  const textColor = !props.withSrsColors ? props.textColor : isHidden ? grey[1] : white[1];
  const bgColor = !props.withSrsColors
    ? props.bgColor
    : isHidden ? grey[6] : SRS_COLORS[getSrsRankName(streak)];

  return {
    vocabId,
    bgColor,
    textColor,
  };
};

export default connect(mapStateToProps)(VocabCard);
