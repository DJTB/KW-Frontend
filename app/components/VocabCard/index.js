import React from 'react';
import PropTypes from 'prop-types';

import * as COLORS from 'shared/styles/colors';

import { Li, Link, Dl } from './styles';

VocabCard.propTypes = {
  id: PropTypes.number.isRequired,
  meaning: PropTypes.string.isRequired,
  readings: PropTypes.array.isRequired,
  color: PropTypes.oneOf(Object.keys(COLORS)),
};

VocabCard.defaultProps = {
  color: 'purple',
};

function VocabCard({ id, meaning, readings, color }) {
  return (
    <Li bgColor={color}>
      <Link plainLink to={`/vocabulary/entry/${id}`}>
        <Dl color={color}>
          <div className="reading">
            {readings.map(({ kana, character }) => ([
              <dt className="kana" lang="ja" >{kana}</dt>,
              <dt className="character" lang="ja" >{character}</dt>,
              <div className="separator" />,
            ]))}
          </div>
          <dd className="meaning">{meaning}</dd>
        </Dl>
      </Link>
    </Li>
  );
}

export default VocabCard;
