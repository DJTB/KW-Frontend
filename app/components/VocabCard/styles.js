import styled from 'styled-components';
import { transparentize, darken } from 'polished';

import * as COLORS from 'shared/styles/colors';
import { ffHeading, mega, gamma, delta } from 'shared/styles/typography';
import { borderRadius } from 'shared/styles/sizing';
import { gutter } from 'shared/styles/layout';

import A from 'base/A';

export const Li = styled.li`
  display: flex;
  flex-direction: column;
  align-items: stretch;
  flex: 1 1 auto;
  line-height: 1;
  margin-right: .3em;
  margin-bottom: .3em;
  background-color: ${({ bgColor }) => COLORS[bgColor]};
  box-shadow: 2px 2px 0 rgba(0,0,0, .1);
  border-radius: ${borderRadius};
  color: ${COLORS.whiteLight};
  font-size: 1.2em;
  max-width: 100%;
  text-decoration: none;
`;

export const Link = styled(A)`
  display: flex;
  flex-direction: column;
  align-content: center;
  justify-content: center;
  align-items: center;
  max-width: 100%;
  color: currentColor;
  flex: 1 0 auto;
`;

export const Dl = styled.dl`
  display: flex;
  margin: 0;
  padding: 1.25em;
  width: 100%;
  flex-flow: row wrap;
  justify-content: center;
  align-content: center;
  flex: 1 0 auto;

  .reading {
    ${gutter()}
    display: flex;
    text-align: center;
    justify-content: center;
    align-items: center;
    align-content: center;
    flex-flow: column;
    flex: 0 1 auto;

    .kana {
      /* no gutter: reading, character, and separator all surround the kana with padding */
      ${delta}
      word-break: keep-all;
      text-shadow: 0 1px 2px ${({ color }) => transparentize(0.3, darken(0.2, COLORS[color]))};
      line-height: 1.1;
    }

    .character {
      ${gutter()}
      ${mega}
      word-break: keep-all;
      text-shadow: 0 2px 2px ${({ color }) => transparentize(0.3, darken(0.2, COLORS[color]))};
    }
  }

  .meaning {
    ${gutter()}
    ${gamma}
    text-transform: capitalize;
    margin: 0;
    flex: 0 1 55%;
    align-self: center;
    text-align: center;
    font-family: ${ffHeading};
    font-weight: 600;
    line-height: 1.25;
    text-shadow: 0 1px 2px ${({ color }) => transparentize(0.7, darken(0.2, COLORS[color]))};
  }

  .separator {
    height: .1rem;
    margin: .2rem 0 .5rem;
    width: 100%;
    background-color: ${({ color }) => transparentize(0.7, darken(0.2, COLORS[color]))};
    border: .075rem solid ${({ color }) => transparentize(0.9, darken(0.2, COLORS[color]))};
    border-radius: 100%;
    &:last-of-type {
      display: none;
    }
  }
`;
