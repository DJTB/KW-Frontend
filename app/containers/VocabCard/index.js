import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose, shouldUpdate } from 'recompose';
import cuid from 'cuid';

import { makeSelectReviewMeanings, makeSelectReviewReadings } from 'shared/selectors';

import * as COLORS from 'shared/styles/colors';
import { Character } from 'components/Reading/styles';
import { Li, Link, Dl } from './styles';

VocabCard.propTypes = {
  color: PropTypes.oneOf(Object.keys(COLORS)),
  id: PropTypes.number.isRequired,
  meanings: PropTypes.array,
  readings: PropTypes.array,
};

VocabCard.defaultProps = {
  color: 'purple',
  meanings: [],
  readings: [],
};

function VocabCard({
  color, id, meanings, readings,
}) {
  return (
    <Li bgColor={color}>
      <Link plainLink to={`/vocabulary/entry/${id}`}>
        <Dl color={color}>
          <div className="reading">
            {readings.map(({ character }) => (
              <Fragment key={cuid()}>
                <Character>{character}</Character>
                <div className="separator" />
              </Fragment>
            ))}
          </div>
          <dd className="meaning">{meanings[0]}</dd>
        </Dl>
      </Link>
    </Li>
  );
}

const mapStateToProps = (state, props) => ({
  meanings: makeSelectReviewMeanings(props.id)(state),
  readings: makeSelectReviewReadings(props.id)(state),
});

export default compose(
  connect(mapStateToProps),
  shouldUpdate((props, nextProps) => props.id !== nextProps.id)
)(VocabCard);
