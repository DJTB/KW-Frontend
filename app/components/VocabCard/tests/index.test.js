import React from 'react';
import { shallow } from 'enzyme';

import VocabCard from '../index';

describe('<VocabCard />', () => {
  const id = 234;
  const meaning = 'Before Long, Soon';
  const readings = [
    {
      character: '近々',
      kana: 'ちかぢか, きんきん, ちかじか',
    },
    {
      character: '間もなく',
      kana: 'まもなく',
    },
  ];

  it('should match baseline snapshot', () => {
    const renderedComponent = shallow(
      <VocabCard
        id={id}
        meaning={meaning}
        readings={readings}
      />
    );
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should adopt color prop', () => {
    const renderedComponent = shallow(
      <VocabCard
        id={id}
        meaning={meaning}
        readings={readings}
        color="orange"
      />
    );
    expect(renderedComponent).toMatchSnapshot();
  });
});
