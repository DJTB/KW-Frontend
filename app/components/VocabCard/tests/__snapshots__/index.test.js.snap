// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VocabCard /> should adopt color prop 1`] = `
<styled.li
  bgColor="orange"
>
  <Styled(A)
    plainLink={true}
    to="/vocabulary/entry/234"
  >
    <styled.dl
      color="orange"
    >
      <div
        className="reading"
      >
        <dt
          className="kana"
          lang="ja"
        >
          ちかぢか, きんきん, ちかじか
        </dt>
        <dt
          className="character"
          lang="ja"
        >
          近々
        </dt>
        <div
          className="separator"
        />
        <dt
          className="kana"
          lang="ja"
        >
          まもなく
        </dt>
        <dt
          className="character"
          lang="ja"
        >
          間もなく
        </dt>
        <div
          className="separator"
        />
      </div>
      <dd
        className="meaning"
      >
        Before Long, Soon
      </dd>
    </styled.dl>
  </Styled(A)>
</styled.li>
`;

exports[`<VocabCard /> should match baseline snapshot 1`] = `
<styled.li
  bgColor="purple"
>
  <Styled(A)
    plainLink={true}
    to="/vocabulary/entry/234"
  >
    <styled.dl
      color="purple"
    >
      <div
        className="reading"
      >
        <dt
          className="kana"
          lang="ja"
        >
          ちかぢか, きんきん, ちかじか
        </dt>
        <dt
          className="character"
          lang="ja"
        >
          近々
        </dt>
        <div
          className="separator"
        />
        <dt
          className="kana"
          lang="ja"
        >
          まもなく
        </dt>
        <dt
          className="character"
          lang="ja"
        >
          間もなく
        </dt>
        <div
          className="separator"
        />
      </div>
      <dd
        className="meaning"
      >
        Before Long, Soon
      </dd>
    </styled.dl>
  </Styled(A)>
</styled.li>
`;
